File: trunk/src/java/org/codehaus/testdox/intellij/actions/ActionEvents.java
package org.codehaus.testdox.intellij.actions;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.DataKeys;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import static jedi.functional.Coercions.array;
import org.codehaus.testdox.intellij.NullPsiElement;
import org.codehaus.testdox.intellij.TestDoxController;
import org.codehaus.testdox.intellij.TestDoxNonJavaFile;
import org.codehaus.testdox.intellij.TestDoxProjectComponent;
import org.codehaus.testdox.intellij.panel.TestDoxToolWindowUI;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

public class ActionEvents {

    static ActionEvents instance = new ActionEvents();

    public TestDoxController getTestDoxController(AnActionEvent event) {
        TestDoxProjectComponent testDoxProjectComponent = getTestDoxProjectComponent(event);
        return (testDoxProjectComponent != null) ? testDoxProjectComponent.getController()
            : Nulls.TESTDOX_CONTROLLER;
    }

    public TestDoxToolWindowUI getTestDoxToolWindowUI(AnActionEvent event) {
        TestDoxProjectComponent testDoxProjectComponent = getTestDoxProjectComponent(event);
        return (testDoxProjectComponent != null) ? testDoxProjectComponent.getTestDoxToolWindowUI()
            : Nulls.TESTDOX_TOOL_WINDOW;
    }

    private TestDoxProjectComponent getTestDoxProjectComponent(AnActionEvent event) {
        return TestDoxProjectComponent.getInstance(event.getData(DataKeys.PROJECT));
    }

    public boolean isJavaFile(AnActionEvent event) {
        TestDoxController testDoxController = getTestDoxController(event);
        VirtualFile file = event.getData(DataKeys.VIRTUAL_FILE);
        return (testDoxController != null) && (file != null) && (testDoxController.getEditorApi().isJavaFile(file));
    }

    public PsiElement getTargetPsiElement(AnActionEvent event) {
        Editor editor = event.getData(DataKeys.EDITOR);
        if (editor == null) {
            return NullPsiElement.INSTANCE();
        }

        PsiFile psiFile = event.getData(DataKeys.PSI_FILE);
        if (psiFile == null) {
            return NullPsiElement.INSTANCE();
        }

        return psiFile.findElementAt(editor.getCaretModel().getOffset());
    }

    static class Nulls {

        static final TestDoxToolWindowUI TESTDOX_TOOL_WINDOW;
        static final TestDoxController TESTDOX_CONTROLLER;

        private static final TestDoxNonJavaFile NON_JAVA_TEST_DOX_FILE = new TestDoxNonJavaFile(null);

        static {
            InvocationHandler invocationHandler = new InvocationHandler() {
                public Object invoke(Object proxy, Method method, Object[] args) {
                    if ("hasActiveEditors".equals(method.getName())) {
                        return false;
                    }
                    if ("canCurrentFileBeUnitTested".equals(method.getName())) {
                        return false;
                    }
                    if (("updatePresentation".equals(method.getName())) && (args[0] instanceof Presentation)) {
                        ((Presentation) args[0]).setEnabled(false);
                    }
                    if ("getCurrentTestDoxFile".equals(method.getName())) {
                        return NON_JAVA_TEST_DOX_FILE;
                    }
                    return null;
                }
            };

            ClassLoader classLoader = invocationHandler.getClass().getClassLoader();
            Class[] interfaces = array(TestDoxToolWindowUI.class, TestDoxController.class);

            Object nullObject = Proxy.newProxyInstance(classLoader, interfaces, invocationHandler);
            TESTDOX_TOOL_WINDOW = (TestDoxToolWindowUI) nullObject;
            TESTDOX_CONTROLLER = (TestDoxController) nullObject;
        }

        private Nulls() {
        }
    }
}

File: src/main/scala/org/codehaus/testdox/intellij/actions/ActionEvents.scala
package org.codehaus.testdox.intellij.actions

import com.intellij.psi.{PsiElement, PsiFile}
import com.intellij.openapi.editor.Editor
import org.codehaus.testdox.intellij.ui.ToolWindowUI
import org.codehaus.testdox.intellij.{TestDoxController, TestDoxNonJavaFile, NullPsiElement, TestDoxProjectComponent}
import java.lang.reflect.{Proxy, Method, InvocationHandler}
import com.intellij.openapi.actionSystem._

class ActionEvents {

  def getTestDoxController(event: AnActionEvent) = {
    val testDoxProjectComponent = getTestDoxProjectComponent(event)
    if (testDoxProjectComponent != null) testDoxProjectComponent.getController else Nulls.TESTDOX_CONTROLLER
  }

  def getToolWindowUI(event: AnActionEvent) = {
    val testDoxProjectComponent = getTestDoxProjectComponent(event)
    if (testDoxProjectComponent != null) testDoxProjectComponent.getToolWindowUI else Nulls.TESTDOX_TOOL_WINDOW
  }

  private def getTestDoxProjectComponent(event: AnActionEvent) = {
    TestDoxProjectComponent.getInstance(event.getData(PlatformDataKeys.PROJECT))
  }

  def isJavaFile(event: AnActionEvent) = {
    val testDoxController = getTestDoxController(event)
    val file = event.getData(PlatformDataKeys.VIRTUAL_FILE)
    testDoxController != null && file != null && testDoxController.getEditorApi.isJavaFile(file)
  }

  def getTargetPsiElement(event: AnActionEvent): PsiElement = {
    val editor: Editor = event.getData(PlatformDataKeys.EDITOR)
    if (editor == null) return NullPsiElement.INSTANCE

    val psiFile: PsiFile = event.getData(LangDataKeys.PSI_FILE)
    if (psiFile == null) return NullPsiElement.INSTANCE

    return psiFile.findElementAt(editor.getCaretModel.getOffset)
  }
}

object ActionEvents {
  var instance = new ActionEvents()
  def setInstance(actionEvents: ActionEvents) { instance = actionEvents }
}

object Nulls {

  private val invocationHandler = new InvocationHandler() {
    private val NON_JAVA_TEST_DOX_FILE = new TestDoxNonJavaFile(null)

    def invoke(proxy: Object, method: Method, args: Array[Object]): Object = {
      if ("hasActiveEditors" == method.getName) return boolean2Boolean(false)
      else if ("canCurrentFileBeUnitTested" == method.getName) return boolean2Boolean(false)
      else if ("update" == method.getName && args(0).isInstanceOf[Presentation]) {
        (args(0).asInstanceOf[Presentation]).setEnabled(false)
      }
      if ("getCurrentTestDoxFile".equals(method.getName)) return NON_JAVA_TEST_DOX_FILE
      return null
    }
  }

  private val classLoader = invocationHandler.getClass.getClassLoader
  private val interfaces = Array(classOf[ToolWindowUI], classOf[TestDoxController])
  private val nullObject = Proxy.newProxyInstance(classLoader, interfaces.asInstanceOf[Array[Class[_]]], invocationHandler)

  val TESTDOX_TOOL_WINDOW = nullObject.asInstanceOf[ToolWindowUI]
  val TESTDOX_CONTROLLER = nullObject.asInstanceOf[TestDoxController]
}
